gh-130902 Add optional socket shutdown to close method of HTTPConnection class in http/client.py

sharktide opened the pull request:
Similar to gh-130862 but for http library

This pr will close #130902 

Overview:
Based on #130902, I added an optional socket shutdown to HTTPConnection.close. The modified function, HTTPConnection.close has a new optional shutdown parameter defaulted to false.

Original function:

```python
  def close(self):
      """Close the connection to the HTTP server."""
      self.__state = _CS_IDLE
      try:
          sock = self.sock
          if sock:
              self.sock = None
              sock.close()   # close it manually... there may be other refs
      finally:
          response = self.__response
          if response:
              self.__response = None
              response.close()
```

Refactored function:

```python
def close(self, shutdown=False):
        self.__state = _CS_IDLE
        try:
            sock = self.sock
            if sock:
                self.sock = None
                if shutdown:
                    try:
                        sock.shutdown(socket.SHUT_RDWR)
                    except OSError as e:
                        print(f"Socket shutdown error: {e}", file=sys.stderr)
                    except Exception as e:
                        print(f"Unexpected error during socket shutdown: {e}", file=sys.stderr)
                try:
                    sock.close()
                except OSError as e:
                    print(f"Socket close error: {e}", file=sys.stderr)
                except Exception as e:
                    print(f"Unexpected error during socket close: {e}", file=sys.stderr)
        finally:
            response = self.__response
            if response:
                self.__response = None
                try:
                    response.close()
                except OSError as e:
                    print(f"Response close error: {e}", file=sys.stderr)
                except Exception as e:
                    print(f"Unexpected error during response close: {e}", file=sys.stderr)
```

This new change has some benefits such as:
- Ensures that all pending data is properly discarded before closing, particularly in scenarios where data might still be buffered.
- Prevents potential issues with lingering resources in some edge cases.
- Aligns with best practices for socket cleanup.
- Can help avoid RST (reset by peer) issues in some network conditions.

View #130902 for more information

I tested it against a custom test script:

```python
import sys
from unittest.mock import MagicMock
from http.client import HTTPConnection  # Replace with the actual module name
from io import StringIO

def test_close_with_shutdown():
    # Create a mock socket object
    mock_socket = MagicMock()
    mock_socket.close = MagicMock()
    mock_socket.shutdown = MagicMock()

    # Instantiate the HTTPConnection object
    connection = HTTPConnection('www.example.com')
    connection.sock = mock_socket

    # Redirect stderr to capture error messages
    original_stderr = sys.stderr
    sys.stderr = StringIO()

    try:
        # Call the close method with shutdown=True
        connection.close(shutdown=True)

        # Check that shutdown and close were called
        mock_socket.shutdown.assert_called_once_with(2)  # SHUT_RDWR
        mock_socket.close.assert_called_once()

        # Check that no error messages were written to stderr
        error_output = sys.stderr.getvalue()
        assert not error_output, f"Error messages were written to stderr: {error_output}"

    finally:
        # Restore original stderr
        sys.stderr = original_stderr

def test_close_without_shutdown():
    # Create a mock socket object
    mock_socket = MagicMock()
    mock_socket.close = MagicMock()
    mock_socket.shutdown = MagicMock()

    # Instantiate the HTTPConnection object
    connection = HTTPConnection('www.example.com')
    connection.sock = mock_socket

    # Redirect stderr to capture error messages
    original_stderr = sys.stderr
    sys.stderr = StringIO()

    try:
        # Call the close method with shutdown=False
        connection.close(shutdown=False)

        # Check that shutdown was not called, but close was
        mock_socket.shutdown.assert_not_called()
        mock_socket.close.assert_called_once()

        # Check that no error messages were written to stderr
        error_output = sys.stderr.getvalue()
        assert not error_output, f"Error messages were written to stderr: {error_output}"

    finally:
        # Restore original stderr
        sys.stderr = original_stderr

def test_close_shutdown_error():
    # Create a mock socket object that raises an OSError on shutdown
    mock_socket = MagicMock()
    mock_socket.close = MagicMock()
    mock_socket.shutdown = MagicMock(side_effect=OSError("Shutdown error"))

    # Instantiate the HTTPConnection object
    connection = HTTPConnection('www.example.com')
    connection.sock = mock_socket

    # Redirect stderr to capture error messages
    original_stderr = sys.stderr
    sys.stderr = StringIO()

    try:
        # Call the close method with shutdown=True
        connection.close(shutdown=True)

        # Check that shutdown and close were called
        mock_socket.shutdown.assert_called_once_with(2)  # SHUT_RDWR
        mock_socket.close.assert_called_once()

        # Check that the error message was written to stderr
        error_output = sys.stderr.getvalue()
        assert "Socket shutdown error: Shutdown error" in error_output, f"Expected error message not found: {error_output}"

    finally:
        # Restore original stderr
        sys.stderr = original_stderr

def test_close_unexpected_error():
    # Create a mock socket object that raises an unexpected exception on shutdown
    mock_socket = MagicMock()
    mock_socket.close = MagicMock()
    mock_socket.shutdown = MagicMock(side_effect=Exception("Unexpected error"))

    # Instantiate the HTTPConnection object
    connection = HTTPConnection('www.example.com')
    connection.sock = mock_socket

    # Redirect stderr to capture error messages
    original_stderr = sys.stderr
    sys.stderr = StringIO()

    try:
        # Call the close method with shutdown=True
        connection.close(shutdown=True)

        # Check that shutdown and close were called
        mock_socket.shutdown.assert_called_once_with(2)  # SHUT_RDWR
        mock_socket.close.assert_called_once()

        # Check that the error message was written to stderr
        error_output = sys.stderr.getvalue()
        assert "Unexpected error during socket shutdown: Unexpected error" in error_output, f"Expected error message not found: {error_output}"

    finally:
        # Restore original stderr
        sys.stderr = original_stderr

if __name__ == '__main__':
    test_close_with_shutdown()
    test_close_without_shutdown()
    test_close_shutdown_error()
    test_close_unexpected_error()
    print("All tests passed.")
```
And it passed successfully:

```shell
PS C:\Users\---\Documents\GitHub\cpython> .\PCbuild\amd64\python_d.exe test.py
All tests passed.
PS C:\Users\---\Documents\GitHub\cpython> 
```


<!-- gh-issue-number: gh-130902 -->
* Issue: gh-130902
<!-- /gh-issue-number -->

sharktide committed:
gh-130902 Add optional socket shutdown to HTTPConnection.close

Similar to gh-130862 but for http library

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

sharktide committed:
Update client.py

sharktide committed:
Update client.py

sharktide committed:
Update client.py

sharktide committed:
Update client.py

Mr-Sunglasses commented on Misc/NEWS.d/next/Library/2025-03-06-16-31-08.gh-issue-130902.mHAtcq.rst:
```diff
@@ -0,0 +1,4 @@
+Issue: :gh:`130902`
+
+
+Modified close(self) method of HTTPConnection class to include an optional socket shutdown similar to gh-130850. Includes error handling. As mentioned, it adds an optional shutdown param to the close method, keeping existing programs working, but adding an extra feature to new programs that offers improvements.
```
- Use a Sphinx reference.
- We generally use present tense in the changelog.

```suggestion
Modifies the :func:`HTTPConnection.close` of :class:`HTTPConnection` to include an optional socket ``shutdown``, improving functionality with error handling.
```

Mr-Sunglasses reviewed (commented):
Some small triage corrections.

Mr-Sunglasses reviewed (commented):
Tests should be added for this.

sharktide committed:
Update Misc/NEWS.d/next/Library/2025-03-06-16-31-08.gh-issue-130902.mHAtcq.rst

Co-authored-by: Kanishk Pachauri <itskanishkp.py@gmail.com>

sharktide committed:
Update test_httplib.py

sharktide committed:
gh-130902 @Mr-Sunglasses removed function and class refs from news entry to avoid docs build error

Mr-Sunglasses commented on Misc/NEWS.d/next/Library/2025-03-06-16-31-08.gh-issue-130902.mHAtcq.rst:
```diff
@@ -0,0 +1,4 @@
+Issue: :gh:`130902`
+
+
+Modifies the ``HTTPConnection.close`` of ``HTTPConnection`` to include an optional socket ``shutdown``, improving functionality with error handling.
```
@sharktide please use this. 
```suggestion
Modifies the :meth:`http.client.HTTPConnection.close` of :class:`http.client.HTTPConnection` to include an optional socket ``shutdown``, improving functionality with error handling.
```

sharktide replied:
kk

sharktide committed:
Update Misc/NEWS.d/next/Library/2025-03-06-16-31-08.gh-issue-130902.mHAtcq.rst

Co-authored-by: Kanishk Pachauri <itskanishkp.py@gmail.com>

Mr-Sunglasses commented on Lib/http/client.py:
```diff
@@ -1012,19 +1012,35 @@ def connect(self):
 
-    def close(self):
-        """Close the connection to the HTTP server."""
+    def close(self, shutdown=False):
```
I think we should add it in the `Doc/library/socket.rst` also, about the `shutdown`.

Ref. https://github.com/python/cpython/blob/b1b4f9625c5f2a6b2c32bc5ee91c9fef3894b5e6/Doc/library/socket.rst?plain=1#L1488

sharktide replied:
ok, shouldn't take too long

sharktide replied:
@Mr-Sunglasses Finally ready for review/ Added/updated docs for the close method of HTTPConnection class as well as added back a modifed docstring in http/client.py!

Mr-Sunglasses reviewed (commented):
- Please update the documentation of `socket.close() ` and mention the use of optional `shutdown`

TiA.

sharktide committed:
gh-130902 @Mr-Sunglasses update documentation for http socket close

sharktide committed:
gh-130902 @Mr-Sunglasses fix typo in update docs

sharktide committed:
Update socket.rst

sharktide committed:
Update socket.rst

sharktide committed:
gh-130902 @Mr-Sunglasses removed trailing whitespace from docs

sharktide committed:
gh-130902 @Mr-Sunglasses add/update docstring in client.py

Updated the docstring of the close method of the HTTPConnection class to reflect optional shutdown parameter in http/client.py

Mr-Sunglasses commented on Doc/library/socket.rst:
```diff
@@ -1485,28 +1485,36 @@ to sockets.
+
+   If the ``shutdown`` parameter is set to ``True``, the socket will first be
+   shut down before closing, ensuring that no further data can be sent or received.
+   This is useful for properly releasing resources and preventing issues like lingering
```
- Just a small suggestion `persistent connection` would be more accurate term, instead of `lingering connection`

```suggestion
   This is useful for properly releasing resources and preventing issues like persistent
```

Mr-Sunglasses reviewed (commented):
Just some small changes in calling of terms.

sharktide committed:
gh-130902 @Mr-Sunglasses update Doc/library/socket.rst

Par suggestion

Co-authored-by: Kanishk Pachauri <itskanishkp.py@gmail.com>

Mr-Sunglasses reviewed (approved):
LGTM, thanks @sharktide

picnixz reviewed (changes requested):
This is incorrect. All exceptions are now swallowed but they should be forwarded so that the caller can handle them how they see fit.

There is no practical need to shutdown the socket. The HTTP server may want to keep the connection alive, possibly to make some cleanup on *their* side. It's just that we don't need it anymore on the *client* side.

Also, there are other places where `.close()` is being called but then without a specific `shutdown` (e.g., `_tunnel()`). Sorry but I'm rejecting this. Also, it's not `socket.close()` that is being changed but `http.client.close()` which is entirely different.
